### ISSUE TITLE
How do I add kernels to kernel_gateway?

### ISSUE THREAD

**Issue by mboratko** [jupyter/kernel_gateway#305]
I've setup a jupyter notebook server and jupyter kernel gateway using conda. I have a number of conda environments setup, and the notebook server correctly displays the full list of notebook servers. I start the kernel gateway as follows:

`jupyter kernelgateway --JupyterWebsocketPersonality.list_kernels=True`

and then browse to `http://localhost:8888/api/kernels`, which reponds with `[]`. I was surprised that it didn't even include the kernel which the kernel gateway is currently running on. How can I specify which kernels the kernel gateway should provide access to?

**Comment by rolweber** [comment]
Did you activate the conda environment of the notebook server before starting the kernel gateway? There's a search path for kernel specs. But iirc, Anaconda also implemented some extra logic to search and enable kernels across conda environments. That logic is not in the kernel gateway by default, and I'm not sure if it will work if you try to configure it.

The kernel gateway is not running on a kernel.

For every type of kernel, there needs to be a directory with a kernelspec in the search path. See here for more information on Jupyter search paths, and how to list them:
https://jupyter.readthedocs.io/en/latest/projects/jupyter-directories.html#id2

I recommend that you list and compare the search paths for your conda notebook server and your kernel gateway server. Maybe it's as simple as setting JUPYTER_DATA_PATH.

**Comment by mboratko** [comment]
I did not, but I was hoping that all available kernels would be available. (I wouldn't want to have to activate each one and then only be able to access that kernel, I was hoping to be able to select from all available kernels on the machine.) That said, there is a base kernel which is active if you have not activated any other environment, and this one wasn't listed either, nor was the default python implementation installed via the OS.

You are correct that conda implements it's own mechanism to expose kernels via the `nb_conda_kernels` package. With this package installed, when I start a notebook normally I see this:

```
/opt/miniconda3/bin/jupyter notebook
[I 12:51:58.082 NotebookApp] [nb_conda_kernels] enabled, 4 kernels found
```

Without this package, however, I still see about 5 kernels, which I believe are found via the [default Jupyter search paths for kernel specs](https://jupyter-client.readthedocs.io/en/latest/kernels.html#kernelspecs). In particular, it picks up the Julia kernel regardless if this package is installed or not. The environment variable `JUPYTER_DATA_PATH` is currently empty. It's possible that `nb_conda_kernels` doesn't expose the conda kernels to the Kernel Gateway, however *with or without* this package I'm not seeing any kernels from the Kernel Gateway.

I also tried manually setting the `JUPYTER_DATA_PATH` environment variable:

```
$ ls /opt/miniconda3/share/jupyter/kernels
python3
$ env JUPYTER_PATH='/opt/miniconda3/share/jupyter/kernels' /opt/miniconda3/bin/jupyter kernelgateway --JupyterWebsocketPersonality.list_kernels=True
```
Same result, navigating to `http://localhost:8888/api/kernels` just returns an empty list `[]`.

**Comment by kevin-bates** [comment]
Can't really comment on python/anaconda environment stuff, but I don't think the `kernels` directory should be part of the `JUPYTER_PATH` value.  Try `export JUPYTER_PATH=/opt/miniconda3/share/jupyter`and restart JKG.  

Tip: If you enable DEBUG logging on JKG (`--log-level=DEBUG`), then a refresh from the Notebook files view will trigger the kernel spec refresh and you'll see the set of kernels that will then appear in the kernels dropdown.  (This is the sequence I use to determine my Notebook is talking to my gateway and the expected kernels are found.)

**Comment by mboratko** [comment]
Thanks for the tip. Here's the output:

```
$ export JUPYTER_PATH=/opt/miniconda3/share/jupyter
$ jupyter kernelgateway --JupyterWebsocketPersonality.list_kernels=True --log-level=DEBUG
[KernelGatewayApp] Searching ['/home/mboratko', '/home/mboratko/.jupyter', '/opt/miniconda3/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[KernelGatewayApp] Looking for jupyter_config in /etc/jupyter
[KernelGatewayApp] Looking for jupyter_config in /usr/local/etc/jupyter
[KernelGatewayApp] Looking for jupyter_config in /opt/miniconda3/etc/jupyter
[KernelGatewayApp] Looking for jupyter_config in /home/mboratko/.jupyter
[KernelGatewayApp] Looking for jupyter_config in /home/mboratko
[KernelGatewayApp] Looking for jupyter_kernel_gateway_config in /etc/jupyter
[KernelGatewayApp] Looking for jupyter_kernel_gateway_config in /usr/local/etc/jupyter
[KernelGatewayApp] Looking for jupyter_kernel_gateway_config in /opt/miniconda3/etc/jupyter
[KernelGatewayApp] Looking for jupyter_kernel_gateway_config in /home/mboratko/.jupyter
[KernelGatewayApp] Loaded config file: /home/mboratko/.jupyter/jupyter_kernel_gateway_config.py
[KernelGatewayApp] Looking for jupyter_kernel_gateway_config in /home/mboratko
[D 180924 16:03:17 selector_events:65] Using selector: EpollSelector
[KernelGatewayApp] Jupyter Kernel Gateway at http://127.0.0.1:8888
```

Visiting `http://localhost:8888/api/kernels` still responds with an empty list, `[]`. I haven't tried connecting a notebook instance elsewhere yet, I assumed visiting this address should respond with a list of kernels so that's what I've been doing in order to test it.

**Comment by rolweber** [comment]
If you just open a terminal and use a Python version from Anaconda, related environment variables are not set. Try `source activate base` before starting the kernel gateway. I'm assuming that you installed the kernel gateway into that base environment.

To find the location of kernel specs across your conda environments, try this:
```bash
find /opt/miniconda3/envs -type d -name kernels
```
Then have a look at the kernel definitions to see if they automatically activate their conda environment, or if that is implemented somewhere in `nb_conda_kernels`. In the latter case, you could write new kernel specs that call a wrapper script which activates the conda environment before launching the kernel.

As to the question why the kernel list is totally empty... there's a config flag to indicate whether Jupyter should put a default kernel (python3) into the list. I had to use it on occasion to prevent Jupyter from adding one when all of my kernelspecs had other names. Maybe that flag is set in your installation.

**Comment by gaocegege** [comment]
I think it may be a bug. I can get the specs from `/api/kernelspecs` but cannot get anything from `/api/kernels`

**Comment by rolweber** [comment]
I'm not that familiar with the API myself... if `/api/kernelspecs` returns the kernel specs, then `/api/kernels` should return the running kernels, right? Did you start a kernel yet?

**Comment by gaocegege** [comment]
@rolweber 

I just start the kernel `jupyter kernelgateway --JupyterWebsocketPersonality.list_kernels=True --log-level=DEBUG`, then visit `http://localhost:8888/api/kernelspecs`, I can get the specs:

```json
{
  "default": "python3",
  "kernelspecs": {
    "ir": {
      "name": "ir",
      "spec": {
        "argv": [
          "/usr/lib/R/bin/R",
          "--slave",
          "-e",
          "IRkernel::main()",
          "--args",
          "{connection_file}"
        ],
        "env": {},
        "display_name": "R",
        "language": "R",
        "interrupt_mode": "signal",
        "metadata": {}
      },
      "resources": {
        "kernel.js": "/kernelspecs/ir/kernel.js",
        "logo-64x64": "/kernelspecs/ir/logo-64x64.png"
      }
    }
    ...
  }
}
```

While I get `[]` when visiting `http://localhost:8888/api/kernels`

It does not affect the usage, I can see the kernels in jupyter notebook, but I just cannot get the result from the API.

**Comment by rolweber** [comment]
_I just start the kernel `jupyter kernelgateway`_

That's not a kernel, that's a _kernel gateway_. A Python process that manages kernels.
Did you tell the kernel gateway to start a kernel? If not, then no kernel is running.

**Comment by gaocegege** [comment]
Oh gotcha, I misunderstand the API, thanks!

**Comment by rolweber** [comment]
Feel free to re-open or to create a new issue if you run into other problems.
